### PR TITLE
Report mismatch segment for strings of unequal lengths

### LINKED ISSUE #907: Show location of difference when comparing two strings
When using `string1.Should().Be(string2)`, if the strings differ it can be awkward to work out where the difference is for large string. E.g. if you're checking that building some json/xml/whatever-long-text works correctly. 

<details>
 <summary>Click for example test output for xml text</summary>

```
Expected xmlString1 to be 
"<CATALOG>
	<CD>
		<TITLE>Empire Burlesque</TITLE>
		<ARTIST>Bob Dylan</ARTIST>
		<COUNTRY>USA</COUNTRY>
		<COMPANY>Columbia</COMPANY>
		<PRICE>10.90</PRICE>
		<YEAR>1985</YEAR>
	</CD>
	<CD>
		<TITLE>Hide your heart</TITLE>
		<ARTIST>Bonnie Tyler</ARTIST>
		<COUNTRY>UK</COUNTRY>
		<COMPANY>CBS Records</COMPANY>
		<PRICE>9.90</PRICE>
		<YEAR>1988</YEAR>
	</CD>
	<CD>
		<TITLE>Greatest Hits</TITLE>
		<ARTIST>Dolly Parton</ARTIST>
		<COUNTRY>USA</COUNTRY>
		<COMPANY>RCA</COMPANY>
		<PRICE>19.90</PRICE>
		<YEAR>1982</YEAR>
	</CD>
	<CD>
		<TITLE>Still got the blues</TITLE>
		<ARTIST>Gary Moore</ARTIST>
		<COUNTRY>UK</COUNTRY>
		<COMPANY>Virgin records</COMPANY>
		<PRICE>10.20</PRICE>
		<YEAR>1990</YEAR>
	</CD>
</CATALOG>" with a length of 756, but 
"<CATALOG>
	<CD>
		<TITLE>Empire Burlesque</TITLE>
		<ARTIST>Bob Dylan</ARTIST>
		<COUNTRY>USA</COUNTRY>
		<COMPANY>Columbia</COMPANY>
		<PRICE>10.90</PRICE>
		<YEAR>1985</YEAR>
	</CD>
	<CD>
		<TITLE>Hide your heart</TITLE>
		<ARTIST>Bonnie Tyler</ARTIST>
		<COUNTRY>UK</COUNTRY>
		<COMPANY>CBS Records</COMPANY>
		<PRICE>9.90</PRICE>
		<YEAR>1988</YEAR>
	</CD>
	<CD>
		<TITLE>Greatest Hits</TITLE>
		<ARTIST>Dolly Parton</ARTIST>
		<COUNTRY>USA</COUNTRY>
		<COMPANY>RCA</COMPANY>
		<PRICE>9.90</PRICE>
		<YEAR>1982</YEAR>
	</CD>
	<CD>
		<TITLE>Still got the blues</TITLE>
		<ARTIST>Gary Moore</ARTIST>
		<COUNTRY>UK</COUNTRY>
		<COMPANY>Virgin records</COMPANY>
		<PRICE>10.20</PRICE>
		<YEAR>1990</YEAR>
	</CD>
</CATALOG>" has a length of 755.
```
</details>

At the moment, I'm copy & pasting from the test output into a diff tool so I can find where the difference is. 

It would be great if the assert message also included a hint to where to the first difference. 

Some ideas that I'd find handy (any of these could go after or before the current message showing the full actual vs expected text)
- For multi-line strings, include the line number of the first difference
- Include a snippet of the string, that shows a few characters before & after the index of the first difference (say 14 characters before the difference, and 6 characters after)
- A [unified diff](https://en.wikipedia.org/wiki/Diff#Unified_format) style display of the two strings (like git would show)

The specific details of how to show what the differences are (or at least the first difference) aren't very important, anything that helps speed up tracking down the differences would be 👍

### PR DESCRIPTION
**Pull request by krajek**
Proposal to fix #907.

I used the same mechanism to display differences as in part of the code that handles strings with equal lengths, with few adjustments. 

BTW Possible follow-ups: 

- "lazy" version of FailWith, that delays the calculation of message and its params
- display differences improvement reporting line within the text when the first difference occurs, maybe only for multiline subjects

### CODE REVIEW

**Comment by dennisdoomen** [comment]
Are you still working on this? If so, please prefix it it with `[WIP]`

**Comment by krajek** [comment]
I think I am done with the first proposition. 
What do you think about the idea of putting a mismatching segment in the assertion message and about this particular implementation?

File: Src/FluentAssertions/Primitives/StringEqualityValidator.cs [hunk 1]
```diff
@@ -27,14 +27,51 @@ protected override bool ValidateAgainstSuperfluousWhitespace()
 
         protected override bool ValidateAgainstLengthDifferences()
         {
+            // Logic is a little bit convoluted because I want to avoid calculation
+            // of mismatch segment in case of equalLength == true for performance reason.
+            // If lazy version of FailWith would be introduced, calculation of mismatch
+            // segment can be moved directly to FailWith's argument
+            bool equalLength = subject.Length == expected.Length;
+
+            string mismatchSegment = GetMismatchSegmentForStringsOfDifferentLengths(equalLength);
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Primitives/StringEqualityValidator.cs]: 🔧 You could prevent that call, but chaining it after the `FailWith` below. See the [CollectionAssertions](https://github.com/fluentassertions/fluentassertions/blob/179fe62b67dfabc4e38c34e9f62afe2d2ead91c8/Src/FluentAssertions/Collections/CollectionAssertions.cs#L1430) for an example

File: Src/FluentAssertions/Primitives/StringEqualityValidator.cs [hunk 2]
```diff
@@ -27,14 +27,51 @@ protected override bool ValidateAgainstSuperfluousWhitespace()
...
+        {
+            if (equalLength)
+            {
+                return "";
+            }
+
+            int indexOfMismatch = subject.IndexOfFirstMismatch(expected, comparisonMode);
+
+            // If there is no difference it means that subject and expected have common prefix
+            // and the first difference is after just that prefix.
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Primitives/StringEqualityValidator.cs]: ❓ What do you mean with _prefix_ here?

File: Src/FluentAssertions/Primitives/StringEqualityValidator.cs [hunk 3]
```diff
@@ -27,14 +27,51 @@ protected override bool ValidateAgainstSuperfluousWhitespace()
...
+                }
+                else
+                {
+                    // If subject is longer we point at first character after expected
+                    indexOfMismatch = Math.Max(0, expected.Length);
+                }
+            }
+
+            string mismatchSegment = subject.IndexedSegmentAt(indexOfMismatch);
+            return mismatchSegment;
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Primitives/StringEqualityValidator.cs]: 🙃  I would just return the resilt of `IndexSegmentAt` directly.

File: Tests/Shared.Specs/BasicEquivalencySpecs.cs [hunk 4]
```diff
@@ -644,7 +644,7 @@ public void
             // Assert
             //-----------------------------------------------------------------------------------------------------------
             act.Should().Throw<XunitException>().WithMessage(
-                @"Expected member Member1 to be*""different"" with a length of 9, but*"""" has a length of 0.*");
+                @"Expected member Member1 to be*""different"" with a length of 9, but*"""" has a length of 0*");
```
Comment (dennisdoomen) [inline Tests/Shared.Specs/BasicEquivalencySpecs.cs]: 🔧  Is the exact exception message important in _all_ those tests? If not, I would only verify it where it's important.

File: Src/FluentAssertions/Primitives/StringEqualityValidator.cs (hunk 1 above)
Comment (krajek) [reply Src/FluentAssertions/Primitives/StringEqualityValidator.cs]: Could you elaborate more on this one? I think I need more guidance. 
Best I could come up with is the code below, but it doesn't even compile.
```
.Given(() => subject)
.ForCondition(x => subject.Length == expected.Length)
.FailWith(
   ExpectationDescription + "{0} with a length of {1}{reason}, but {2} has a length of {3}, differs near {4}.",
   (_) => expected, (_) => expected.Length, (x) => x, (x) => x.Length, (_) => mismatchSegment)
 .SourceSucceeded; // this line fails 
```
Even if the last list compiled it still feels like a "workaround" for the fact that "lazy" version of FailWith is not available on ```AssertionScope```, doesn't it?

Comment (dennisdoomen) [reply Src/FluentAssertions/Primitives/StringEqualityValidator.cs]: Yeah, I got your approach wrong. You want to combine everything in one failure message.

File: Src/FluentAssertions/Primitives/StringEqualityValidator.cs [hunk 5]
```diff
@@ -27,14 +27,56 @@ protected override bool ValidateAgainstSuperfluousWhitespace()
...
+                    // We would like to point at next character as it is the real
+                    // index of first mismatch, but we need to point at character existing in
+                    // subject, so the next best thing is the last subject character.
+                    indexOfMismatch = Math.Max(0, subject.Length - 1);
+                }
+                else
+                {
+                    // If subject is longer we are sure that subject starts with expected
+                    // and we point at first character after expected.
+                    indexOfMismatch = Math.Max(0, expected.Length);
```
Comment (jnyrup) [inline Src/FluentAssertions/Primitives/StringEqualityValidator.cs]: 🔧 `Math.Max` is not be needed as `expected.Lengh` is `>= 0`.

File: Src/FluentAssertions/Primitives/StringEqualityValidator.cs [hunk 6]
```diff
@@ -27,14 +27,56 @@ protected override bool ValidateAgainstSuperfluousWhitespace()
...
                 .FailWith(
-                    ExpectationDescription + "{0} with a length of {1}{reason}, but {2} has a length of {3}.",
+                    ExpectationDescription + "{0} with a length of {1}{reason}, but {2} has a length of {3}, differs near " + mismatchSegment +  ".",
                     expected, expected.Length, subject, subject.Length)
                 .SourceSucceeded;
         }
 
+        private string GetMismatchSegmentForStringsOfDifferentLengths(bool equalLength)
+        {
+            if (equalLength)
```
Comment (jnyrup) [inline Src/FluentAssertions/Primitives/StringEqualityValidator.cs]: 🔧 A special reason why `equalLength` is a parameter, when both `subject` and `expected` are available in the method scope?

**Comment by dennisdoomen** [comment]
@krajek can we merge this?

File: Src/FluentAssertions/Primitives/StringEqualityValidator.cs (hunk 5 above)
Comment (krajek) [reply Src/FluentAssertions/Primitives/StringEqualityValidator.cs]: Good catch, thanks.

**Comment by krajek** [comment]
Yes, please. Thanks for the reviews.
After this one is merged I will try to follow up will lazy `FailWith` and refactor accordingly.

**Comment by dennisdoomen** [comment]
Awesome. Thanks man.

**Pull request merged**